app/policies/order/delivery_status_policy.rb

```diff
@@ -0,0 +1,23 @@
+class Order::DeliveryStatusPolicy
+
+  def initialize(user, order)
+    @user = user
+    @order = order
+  end
+
+  def mark_delivered?
+    !user.buyer_only? && order.undelivered_for_user?(user) &&
+      (user.can_manage_market?(order.market) || order.delivery.delivery_schedule.direct_to_customer?)
+  end
+
+  def mark_undelivered?
+    (order.market.is_buysell_market? && !order.undelivered_for_user?(user) && user.can_manage_market?(order.market)) ||
+      (order.market.is_consignment_market? && user.can_manage_market?(order.market) &&
+        (order.delivery_status == 'pending' || order.delivery_status == 'partially delivered') && Inventory::Utils.consignment_can_undeliver?(order))
+  end
```

Comment:
Multiple delegate policies would make it easier to delete the consignment stuff later.

Reply:
Looks like you could extract a guard here for both regular and consignment markets:

````
def mark_undelivered?
    return false if order.undelivered_for_user?(user)
````

Reply:
I actually have a mistake in the consignment part as `pending` should be `delivered`. So I'm not sure I can extract a guard for the consignment case, can I? 

I did explore that a bit, but got scared off that `undelivered_for_user?` looks through order_items but this original code just references the `order`'s delivery_status. Maybe we can discuss...

Reply:
So are you saying I should have two delegates for `mark_undelivered?`? One for buysell one for consignment?

Reply:
Could you send me that code again? I squashed my PR prematurely and lost the WIP commit you put in that branch.

Reply:
Something like this?
```
  def market_undelivered?
    if order.market.is_buysell_market?
      DeliveryStatusDeliveredPolicy.new(current_user, order).mark_undelivered?
    else
      Consignment::DeliveryStatusDeliveredPolicy.new(current_user, order).mark_undelivered?
    end
  end